<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>ClassifyGxT - classifying gene-by-treatment interactions • classifygxt</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="ClassifyGxT - classifying gene-by-treatment interactions">
<meta property="og:description" content="Classify gene-by-treatment (GxT) interactions, taking a list of SNPs with significant GxT as input and returning posterior probabilities of GxT types.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">classifygxt</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/classifygxt.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="classifygxt">ClassifyGxT<a class="anchor" aria-label="anchor" href="#classifygxt"></a>
</h1></div>
<div class="section level2">
<h2 id="about">About<a class="anchor" aria-label="anchor" href="#about"></a>
</h2>
<p>ClassifyGxT is a method for classifying gene-by-treatment (GxT) interactions using Bayesian model selection (BMS). The method is primarily designed for molecular count phenotypes, such as gene expression and chromatin accessibility, although it can be used for other types of phenotypes. It takes as input a list of feature-SNP pairs (or SNPs) with significant GxT interactions that have already been identified by a standard method and assigns posterior probability of different types of GxT interactions to each of the feature-SNP pairs (or SNPs). See <strong>Overview</strong> below for the types of GxT interactions.</p>
<p>See <code>Get started</code> and <code>Reference</code> for usage instructions.</p>
</div>
<div class="section level2">
<h2 id="publication">Publication<a class="anchor" aria-label="anchor" href="#publication"></a>
</h2>
<p>Yuriko Harigaya, Nana Matoba, Brandon D. Le Jordan M. Valone Jason L. Stein Michael I. Love* William Valdar*. “Probabilistic classification of gene-by-treatment interactions on molecular count phenotypes.” doi: <a href="https://doi.org/10.1101/2024.08.03.605142" class="external-link uri">https://doi.org/10.1101/2024.08.03.605142</a> (* These authors contributed equally to this work.)</p>
</div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>In the simplest form, our BMS framework can be described as follows. For <span class="math inline"><em>n</em></span> samples indexed by <span class="math inline"><em>i</em></span>, we consider a linear model shown below. <span class="math inline"><em>y</em><sub><em>i</em></sub></span> is the phenotype, <span class="math inline"><em>g</em><sub><em>i</em></sub></span> is the genotype, and <span class="math inline"><em>t</em><sub><em>i</em></sub></span> is the indicator variable as to whether the sample <span class="math inline"><em>i</em></span> was treated. <span class="math inline"><em>ε</em><sub><em>i</em></sub></span> is the residual error (noise) and is assumed to follow a Normal distributiion with a zero mean and variance <span class="math inline"><em>σ</em><sup>2</sup></span>. <span class="math inline"><em>β</em><sub>0</sub></span> is the intercept. <span class="math inline"><em>β</em><sub><em>g</em></sub></span>, <span class="math inline"><em>β</em><sub><em>t</em></sub></span>, and <span class="math inline"><em>β</em><sub><em>g</em> × <em>t</em></sub></span> are the genotype, treatment, and GxT interaction effects, respectively. We use the <span class="math inline"><em>m</em><sub><em>g</em></sub></span>, <span class="math inline"><em>m</em><sub><em>t</em></sub></span>, and <span class="math inline"><em>m</em><sub><em>g</em> × <em>t</em></sub></span> indicator variables to specify the inclusion and exclusion of the genotype, treatment, and GxT interaction terms, respectively. The <span class="math inline"><strong>m</strong><sub><em>j</em></sub></span> vector refers to the <span class="math inline"><em>j</em></span>-th model (<span class="math inline"><em>j</em> = 1, …, 8</span>). The eight models represent different types of GxT interactions.</p>
<center>
<img src="reference/figures/bms.png" align="center" width="600">
</center>
<p>To account for the inherent nonlinear relationship between the genotype and the tranformed molecular count phenotype, we use nonlinear regression. The eight models can be formulated in the same way using the <span class="math inline"><strong>m</strong></span> vectors. See our manuscript (<strong>Publication</strong>) for prior specification, posterior computation, and other details.</p>
<p>Although we primarily use the <span class="math inline"><strong>m</strong></span> vector notations to specify the types of GxT interaction, it is also useful to consider the following nomenclatures.</p>
<table class="table">
<thead><tr class="header">
<th>
<span class="math inline"><strong>m</strong></span> vector</th>
<th>Nomenclature</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>(0,0,0)</td>
<td>No effect</td>
</tr>
<tr class="even">
<td>(1,0,0)</td>
<td>Genotype main effect only</td>
</tr>
<tr class="odd">
<td>(0,1,0)</td>
<td>Treatment main effect only</td>
</tr>
<tr class="even">
<td>(1,1,0)</td>
<td>Genotype and treatment main effects only</td>
</tr>
<tr class="odd">
<td>(0,0,1)</td>
<td>Treatment-induced genotype effect</td>
</tr>
<tr class="even">
<td>(1,0,1)</td>
<td>Treatment-altered genotype effect</td>
</tr>
<tr class="odd">
<td>(0,1,1)</td>
<td>Treatment main effect and treatment-induced genotype effect</td>
</tr>
<tr class="even">
<td>(1,1,1)</td>
<td>Genotype and treatment main effects and interaction</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The <em>classifygxt</em> R package can be installed using <em>devtools</em>.</p>
<pre><code><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"devtools"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"yharigaya/classifygxt"</span><span class="op">)</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing classifygxt</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Yuriko Harigaya <br><small class="roles"> Author, maintainer </small>  </li>
<li>Michael Love <br><small class="roles"> Author </small>  </li>
<li>William Valdar <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Yuriko Harigaya, Michael Love, William Valdar.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
