<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>ClassifyGxT - classifying gene-by-treatment interactions • classifygxt</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="ClassifyGxT - classifying gene-by-treatment interactions">
<meta property="og:description" content="classifygxt">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">classifygxt</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/classifygxt.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/tensorqtl.html">Using ClassifyGxT with TensorQTL</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/yharigaya/classifygxt/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>ClassifyGxT - classifying gene-by-treatment
interactions</h1>
                        <h4 data-toc-skip class="author">Yuriko
Harigaya, Michael Love, William Valdar</h4>
            
            <h4 data-toc-skip class="date">2024-11-22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/yharigaya/classifygxt/blob/HEAD/vignettes/classifygxt.Rmd" class="external-link"><code>vignettes/classifygxt.Rmd</code></a></small>
      <div class="hidden name"><code>classifygxt.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The ClassifyGxT software is an implementation of a Bayesian model
selection (BMS) framework for classifying GxT interactions. The method
was developed primarily for molecular count phenotypes, such as RNA-seq
and ATAC-seq data, although it can be used for other types of
phenotypes.</p>
<ul>
<li>The input is a list of feature-SNP pairs for which significant GxT
interactions have been identified by the standard response molecular QTL
mapping procedure (see <span class="citation">HIPSCI Consortium et al.
(2018)</span> and <span class="citation">Matoba et al. (2024)</span> for
example). For each of the feature-SNP pairs, individual genotype and
phenotype data are required.</li>
<li>The main output is the posterior probability for different GxT
types. See <a href="https://yharigaya.github.io/classifygxt/#overview" class="external-link">Overview</a> on
the main page for the model categories representing the types of GxT
interactions.</li>
<li>The package provides functions to process a single feature-SNP pair.
In practice, we recommend running parallel jobs on the computer cluster
with appropriate grouping.</li>
</ul>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>We first show the minimal steps to run ClassifyGxT. We assume that
<code>data</code> is a list object, as described in <a href="https://yharigaya.github.io/classifygxt/articles/classifygxt.html#input-data" class="external-link">Input
data</a>. See <a href="https://yharigaya.github.io/classifygxt/articles/classifygxt.html#performing-bms" class="external-link">Performing
BMS</a> for setting hyperparameter values.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set hyperparameter values</span></span>
<span><span class="va">phi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.5</span>, <span class="fl">2.0</span>, <span class="fl">1.0</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># perform bms</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/do_bms.html">do_bms</a></span><span class="op">(</span></span>
<span>    data<span class="op">=</span><span class="va">data</span>, phi<span class="op">=</span><span class="va">phi</span>,</span>
<span>    fn<span class="op">=</span><span class="st">"nonlinear"</span>, method<span class="op">=</span><span class="st">"map.lap"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract posterior probabilities</span></span>
<span><span class="va">pp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_pp.html">get_pp</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h2>
<p>For each feature-SNP pair, a list object containing the following
elements needs to be generated:</p>
<ul>
<li>
<code>y</code>: A vector of length <span class="math inline">\(n\)</span> containing preprocessed molecular count
phenotypes, where <span class="math inline">\(n\)</span> is the number
of samples. See <a href="https://yharigaya.github.io/classifygxt/articles/classifygxt.html#data-preprocessing" class="external-link">Data
preprocessing</a> below for preprocessing.</li>
<li>
<code>g</code>: A vector of length <span class="math inline">\(n\)</span> containing genotypes coded as <span class="math inline">\({0, 1, 2}\)</span> to represent the number of
minor (alternative) alleles or the imputation-based allelic dosage in
<span class="math inline">\([0, 2]\)</span>.</li>
<li>
<code>t</code>: A vector of indicator variables for the
treatment.</li>
<li>
<code>subject</code>: An optional character or numeric vector
corresponding to the subjects. This is necessary when the model includes
donor or polygenic (kinship) random effects.</li>
<li>
<code>feat.id</code>: An optional character string representing the
feature.</li>
<li>
<code>snp.id</code>: An optional character string representing the
SNP.</li>
</ul>
<p>Note that the samples must be in the same order in the
<code>y</code>, <code>g</code>, <code>t</code>, and <code>subject</code>
elements and that the list can contain additional elements.</p>
</div>
<div class="section level2">
<h2 id="data-preprocessing">Data preprocessing<a class="anchor" aria-label="anchor" href="#data-preprocessing"></a>
</h2>
<p>For molecular count phenotypes, the raw count data from sequencing
experiments can be processed as follows:</p>
<ul>
<li>Step 1: make a matrix of feature counts with rows and columns
representing features and samples, respectively</li>
<li>Step 2: scale the count data by the library size according to <span class="citation">Palowitch et al. (2018)</span>
</li>
<li>Step 3: optionally filter features based on the scaled counts
according to <span class="citation">Matoba et al. (2024)</span>
</li>
<li>Step 4: transform using the function <span class="math inline">\(\log(x + 1)\)</span>
</li>
<li>Step 5: regress out the treatment indicator (and optionally other
fixed-effect covariates, such as sex and age of the subjects)</li>
<li>Step 6: perform principal component analysis (PCA)</li>
<li>Step 7: regress out an appropriate number of PCs (and the same set
of fixed-effect covariates as in step 5) from the original, transformed
data (from step 4)</li>
</ul>
<p>Note that ClassifyGxT currently does not accommodate covariates other
than the donor or polygenic random effects. Other covariates need to be
regressed out prior to performing BMS (step 7). See <a href="https://yharigaya.github.io/classifygxt/articles/classifygxt.html#including-covariates" class="external-link">Including
covariates</a> for details. For determining the number of PCs, see <span class="citation">Matoba et al. (2024)</span>.</p>
</div>
<div class="section level2">
<h2 id="loading-packages">Loading packages<a class="anchor" aria-label="anchor" href="#loading-packages"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/yharigaya/classifygxt" class="external-link">classifygxt</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="generating-data">Generating data<a class="anchor" aria-label="anchor" href="#generating-data"></a>
</h2>
<p>In this section, we simulate data using <code><a href="../reference/make_data.html">make_data()</a></code>. The
simulated data will have the same format as discussed in the previous
section and will be used for demonstrating how to run BMS in the next
sections. For simplicity, we omit random effects. See the function
documentation (<code><a href="../reference/make_data.html">?make_data</a></code>) for how to include random
effects.</p>
<p>We first specify the number of feature-SNP pairs for each of the
eight models.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">model.name</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_model_names.html">get_model_names</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "0,0,0" "1,0,0" "0,1,0" "1,1,0" "0,0,1" "1,0,1" "0,1,1" "1,1,1"</span></span></code></pre></div>
<p>For this demonstration, we only make data for 10 pairs for each
model.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">num</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">8</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">num</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">model.name</span></span>
<span><span class="va">num</span></span>
<span><span class="co">#&gt; 0,0,0 1,0,0 0,1,0 1,1,0 0,0,1 1,0,1 0,1,1 1,1,1 </span></span>
<span><span class="co">#&gt;    10    10    10    10    10    10    10    10</span></span></code></pre></div>
<p>With this specification, the output will be a list of 80 lists. Each
of the lists corresponds to each feature-SNP pair. The first ten lists
are based on <code>"0,0,0"</code>, the next ten are based on
<code>"1,0,0"</code>, and so forth.</p>
<p>The genotype, treatment, and interaction effects will be drawn from
Normal distributions with a zero mean and user-specified standard
deviations. These values represent “typical” magnitudes of the effects.
We specify standard deviations of the effects as follows.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sd.g</span> <span class="op">&lt;-</span> <span class="fl">1.5</span> <span class="co"># genotype</span></span>
<span><span class="va">sd.t</span> <span class="op">&lt;-</span> <span class="fl">2.0</span> <span class="co"># treatment</span></span>
<span><span class="va">sd.gxt</span> <span class="op">&lt;-</span> <span class="fl">1.0</span> <span class="co"># interaction</span></span>
<span><span class="va">sd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">sd.g</span>, <span class="va">sd.t</span>, <span class="va">sd.gxt</span><span class="op">)</span></span></code></pre></div>
<p>Note that the residual error standard deviation <span class="math inline">\(\sigma\)</span>, which represents a typical
magnitude of noise, is set to 1 by default.</p>
<p>The following code generates a data frame specifying the mapping
between samples, subjects, and treatment conditions.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n.sub</span> <span class="op">&lt;-</span> <span class="fl">80</span> <span class="co"># number of subjects</span></span>
<span><span class="va">anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    subject<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">n.sub</span><span class="op">)</span>, each<span class="op">=</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>    condition<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, times<span class="op">=</span><span class="va">n.sub</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">anno</span><span class="op">)</span></span>
<span><span class="co">#&gt;   subject condition</span></span>
<span><span class="co">#&gt; 1       1         0</span></span>
<span><span class="co">#&gt; 2       1         1</span></span>
<span><span class="co">#&gt; 3       2         0</span></span>
<span><span class="co">#&gt; 4       2         1</span></span>
<span><span class="co">#&gt; 5       3         0</span></span>
<span><span class="co">#&gt; 6       3         1</span></span></code></pre></div>
<p>Now we can generate fake data using <code><a href="../reference/make_data.html">make_data()</a></code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_data.html">make_data</a></span><span class="op">(</span></span>
<span>    anno<span class="op">=</span><span class="va">anno</span>, fn<span class="op">=</span><span class="st">"nonlinear"</span>,</span>
<span>    num<span class="op">=</span><span class="va">num</span>, sd<span class="op">=</span><span class="va">sd</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="performing-bms">Performing BMS<a class="anchor" aria-label="anchor" href="#performing-bms"></a>
</h2>
<p>In this section, we perform BMS using <code><a href="../reference/do_bms.html">do_bms()</a></code>. For
simplicity, we do not model random effects. See <a href="https://yharigaya.github.io/classifygxt/articles/classifygxt.html#including-random-effects" class="external-link">Including
random effects</a> below for how to include random effects.</p>
<p>We first specify the “model prior.” Note that this is an optional
argument and that, by default, a uniform prior is used. This
specification reflects a prior belief that all models are equally
likely. Here, we explicitly specify the default model prior for
demonstration purposes.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p.m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">8</span>, <span class="fl">8</span><span class="op">)</span></span></code></pre></div>
<p>We also specify the “effect prior” by choosing the values of the
<span class="math inline">\(\phi_g\)</span>, <span class="math inline">\(\phi_t\)</span>, and <span class="math inline">\(\phi_{g \times t}\)</span> hyperparameters, which
correspond to the genotype, treatment, and GxT interaction effects,
respectively. The hyperparameters represent our prior beliefs about the
effects relative to the residual error standard deviation (noise).
Specifically, we place priors, <span class="math display">\[\begin{equation}
  \beta_g \sim \mathrm{N}(0, \phi_g^2 \sigma^2), \quad
  \beta_t \sim \mathrm{N}(0, \phi_t^2 \sigma^2), \quad
  \beta_{g \times t} \sim \mathrm{N}(0, \phi_{g \times t}^2 \sigma^2).
\end{equation}\]</span> In practice, we recommend optimizing the values
via an empirical Bayes approach (see <a href="https://yharigaya.github.io/classifygxt/articles/classifygxt.html#optimizing-the-effect-prior-hyperparameters" class="external-link">Optimizing
the effect prior hyperparameters</a> below). Here, we set them to the
same values as the effect standard deviations in the data-generating
models. This is a reasonable choice since we set the residual error
standard deviation to 1 when generating the data.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phi.g</span> <span class="op">&lt;-</span> <span class="fl">1.5</span> <span class="co"># genotype</span></span>
<span><span class="va">phi.t</span> <span class="op">&lt;-</span> <span class="fl">2.0</span> <span class="co"># treatment</span></span>
<span><span class="va">phi.gxt</span> <span class="op">&lt;-</span> <span class="fl">1.0</span> <span class="co"># interaction</span></span>
<span><span class="va">phi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">phi.g</span>, <span class="va">phi.t</span>, <span class="va">phi.gxt</span><span class="op">)</span></span></code></pre></div>
<p>The following code performs BMS for the 71st data using
<code><a href="../reference/do_bms.html">do_bms()</a></code> with nonlinear regression
(<code>fn="nonlinear"</code>). Recall that the 71-80th data were
generated based on the eighth model, <code>"1,1,1"</code> (see <a href="https://yharigaya.github.io/classifygxt/articles/classifygxt.html#generating-data" class="external-link">Generating
data</a> above). The <code>method</code> input argument must be set to
either <code>"mcmc.bs"</code>, which represents Markov Chain Monte Carlo
(MCMC) followed by bridge sampling, or <code>"map.lap"</code>, which
represents MAP estimation followed by Laplace approximation. Although
the latter method can be orders of magnitude faster, we recommend using
<code>"mcmc.bs"</code> for the final results, if possible, since it is
not straightforward to obtain error bounds for Laplace
approximation.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k</span> <span class="op">&lt;-</span> <span class="fl">71</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data.list</span><span class="op">[[</span><span class="va">k</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/do_bms.html">do_bms</a></span><span class="op">(</span></span>
<span>    data<span class="op">=</span><span class="va">data</span>, p.m<span class="op">=</span><span class="va">p.m</span>, phi<span class="op">=</span><span class="va">phi</span>,</span>
<span>    fn<span class="op">=</span><span class="st">"nonlinear"</span>, method<span class="op">=</span><span class="st">"map.lap"</span><span class="op">)</span></span></code></pre></div>
<p>This returns a list object containing the following elements.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "fn"             "ranef"          "rint"           "p.m"           </span></span>
<span><span class="co">#&gt; [5] "seed"           "ln.p.y.given.m" "ln.p.y"         "p.m.given.y"   </span></span>
<span><span class="co">#&gt; [9] "optim.list"</span></span></code></pre></div>
<ul>
<li>
<code>fn</code> - A character string specifying the function.</li>
<li>
<code>ranef</code> - A logical.</li>
<li>
<code>rint</code> - A logical as to whether the phenotypes have been
RINT-transformed.</li>
<li>
<code>p.m</code> - A vector of hyperparameters of the model
prior.</li>
<li>
<code>seed</code> - A integer specifying a seed fo RNG.</li>
<li>
<code>ln.p.y.given.m</code> - A named vector of the log marginal
likelihood given each model. If method is set to <code>"map.lap"</code>,
the value is shifted by a constant.</li>
<li>
<code>ln.p.y</code> - A scalar value of the log marginal likelihood.
If method is set to <code>"map.lap"</code>, the value is shifted by a
constant.</li>
<li>
<code>p.m.given.y</code> - A named vector of posterior probability
of the models.</li>
<li>
<code>optim.list</code> - A list of outputs from the optim function
from the stat package containing MAP estimates and Hessian. This element
is included only when method is set to <code>"map.lap"</code>.</li>
</ul>
<p>See the function documentation (<code><a href="../reference/do_bms.html">?do_bms</a></code>) for more
details.</p>
<p>For multiple feature-SNP pairs, we recommend processing the data in
batches using a workflow such as <a href="https://snakemake.readthedocs.io/en/stable/" class="external-link">Snakemake</a>. Within
a batch, BMS can be performed in serial or parallel processes. For a
moderate number of feature-SNP pairs, the following code can be
used.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span></span>
<span>    X<span class="op">=</span><span class="va">data.list</span>, FUN<span class="op">=</span><span class="va">do_bms</span>, p.m<span class="op">=</span><span class="va">p.m</span>, phi<span class="op">=</span><span class="va">phi</span>,</span>
<span>    fn<span class="op">=</span><span class="st">"nonlinear"</span>, method<span class="op">=</span><span class="st">"map.lap"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="extracting-the-posterior-probability-of-the-models">Extracting the posterior probability of the models<a class="anchor" aria-label="anchor" href="#extracting-the-posterior-probability-of-the-models"></a>
</h3>
<p>We can extract the posterior probability of the model as follows.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">pp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_pp.html">get_pp</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;        0,0,0        1,0,0        0,1,0        1,1,0        0,0,1        1,0,1 </span></span>
<span><span class="co">#&gt; 2.381904e-31 3.102070e-13 1.859044e-25 6.219652e-01 6.612267e-19 5.423350e-12 </span></span>
<span><span class="co">#&gt;        0,1,1        1,1,1 </span></span>
<span><span class="co">#&gt; 3.940914e-14 3.780348e-01</span></span></code></pre></div>
<p>Note that, in this particular example, the highest posterior
probability is assigned to the fourth model <code>"1,1,0"</code>, even
though the data-generating model is “<code>1,1,1</code>”. However, we
will see that BMS tends to choose the correct models across multiple
feature-SNP pairs (<a href="https://yharigaya.github.io/classifygxt/articles/classifygxt.html#heatmap-of-the-posterior-probability-of-the-models" class="external-link">Heatmap
of the posterior probability of the models</a> below).</p>
<p>The following code can be used to extract the posterior probability
from a list object storing results for multiple feature-SNP pairs.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pp.mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span>X<span class="op">=</span><span class="va">res.list</span>, FUN<span class="op">=</span><span class="va">get_pp</span><span class="op">)</span></span></code></pre></div>
<p>In this case, the output is a matrix that contains rows and columns
corresponding to the models and the feature-SNP pairs, respectively.</p>
</div>
</div>
<div class="section level2">
<h2 id="including-covariates">Including covariates<a class="anchor" aria-label="anchor" href="#including-covariates"></a>
</h2>
<div class="section level3">
<h3 id="including-fixed-effect-covariates">Including fixed-effect covariates<a class="anchor" aria-label="anchor" href="#including-fixed-effect-covariates"></a>
</h3>
<p>In molecular QTL analyses on experimental data, it is important to
control for confounding factors, such as molecular phenotype PCs and
PEER factors to avoid spurious associations. We recommend identifying
PCs according to the procedure described in Data preprocessing (steps 1
- 6). For computational reasons, ClassifyGxT does not currently
accommodate covariates other than the donor or polygenic random effects.
The fixed-effect covariates need to be regressed out prior to performing
BMS as in <span class="citation">Matoba et al. (2024)</span>. See step 7
in <a href="https://yharigaya.github.io/classifygxt/articles/classifygxt.html#data-preprocessing" class="external-link">Data
preprocessing</a>. Note that we do not typically consider fixed-effect
covariates in simulation experiments.</p>
</div>
<div class="section level3">
<h3 id="including-random-effects">Including random effects<a class="anchor" aria-label="anchor" href="#including-random-effects"></a>
</h3>
<p>In analyses of experimental data, it is ofen desirable to include
random effects in the model. To include donor random effects, we first
need to run <code><a href="../reference/get_tu_lambda.html">get_tu_lambda()</a></code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tu.lambda</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_tu_lambda.html">get_tu_lambda</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<p>We then use the output as input when running
<code><a href="../reference/do_bms.html">do_bms()</a></code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res.ranef</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/do_bms.html">do_bms</a></span><span class="op">(</span></span>
<span>    data<span class="op">=</span><span class="va">data</span>, p.m<span class="op">=</span><span class="va">p.m</span>, phi<span class="op">=</span><span class="va">phi</span>,</span>
<span>    fn<span class="op">=</span><span class="st">"nonlinear"</span>, method<span class="op">=</span><span class="st">"map.lap"</span>,</span>
<span>    tu.lambda<span class="op">=</span><span class="va">tu.lambda</span><span class="op">)</span></span></code></pre></div>
<p>We can also include polygenic random effects rather than donor random
effects to accout for the genetic relatedness and popuplation structure.
See the function documentation (<code><a href="../reference/get_tu_lambda.html">?get_tu_lambda</a></code>) for
details.</p>
</div>
</div>
<div class="section level2">
<h2 id="visualizing-the-results">Visualizing the results<a class="anchor" aria-label="anchor" href="#visualizing-the-results"></a>
</h2>
<div class="section level3">
<h3 id="generating-a-genotype-phenotype-gp-plot">Generating a genotype-phenotype (GP) plot<a class="anchor" aria-label="anchor" href="#generating-a-genotype-phenotype-gp-plot"></a>
</h3>
<p>We recommend visually inspecting the model fit by generating scatter
plots, which we call a “GP” plot, for each feature-SNP pair (or SNP). To
make a GP plot, we first create a list of data frames using
<code><a href="../reference/format_gp.html">format_gp()</a></code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gp.plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/format_gp.html">format_gp</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">data</span>, fit<span class="op">=</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<p>We then create a <code>ggplot2</code> object using
<code><a href="../reference/make_gp_plot.html">make_gp_plot()</a></code>. The appearance of the plot can be modified
as usual.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_gp_plot.html">make_gp_plot</a></span><span class="op">(</span>gp<span class="op">=</span><span class="va">gp.plot</span><span class="op">)</span></span>
<span><span class="va">p1</span></span></code></pre></div>
<p><img src="classifygxt_files/figure-html/unnamed-chunk-18-1.png" width="384" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="generating-a-posterior-probability-pp-plot">Generating a posterior probability (PP) plot<a class="anchor" aria-label="anchor" href="#generating-a-posterior-probability-pp-plot"></a>
</h3>
<p>It is also useful to generate a barplot of posterior probability,
which we call a “PP” plot. To make a PP plot, we first create a data
frame using <code><a href="../reference/format_pp.html">format_pp()</a></code>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pp.plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/format_pp.html">format_pp</a></span><span class="op">(</span>fit<span class="op">=</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<p>We then create a <code>ggplot2</code> object using
<code><a href="../reference/make_pp_plot.html">make_pp_plot()</a></code>. The appearance of the plot can be modified
as usual.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_pp_plot.html">make_pp_plot</a></span><span class="op">(</span>pp<span class="op">=</span><span class="va">pp.plot</span><span class="op">)</span> </span>
<span><span class="va">p2</span></span></code></pre></div>
<p><img src="classifygxt_files/figure-html/unnamed-chunk-20-1.png" width="384" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="heatmap-of-the-posterior-probability-of-the-models">Heatmap of the posterior probability of the models<a class="anchor" aria-label="anchor" href="#heatmap-of-the-posterior-probability-of-the-models"></a>
</h3>
<p>We can generate a heatmap to visualize posterior probability across
multiple feature-SNP pairs using <code><a href="../reference/make_heatmap.html">make_heatmap()</a></code>. Note that
we transpose the matrix using <code><a href="https://rdrr.io/r/base/t.html" class="external-link">t()</a></code> in the following
code.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_heatmap.html">make_heatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">pp.mat</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">p3</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"bottom"</span>,</span>
<span>           legend.title<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="classifygxt_files/figure-html/unnamed-chunk-21-1.png" width="480" style="display: block; margin: auto;"></p>
<p>As expected, we see that the highest probability tends to be assigned
to the correct (i.e., data-generating) model. That is, the MAP model
tends to be <code>"0,0,0"</code> for the first ten feature-SNP pairs in
the left-most columns, <code>"1,0,0"</code> for the 11-20th pairs, and
so forth.</p>
</div>
</div>
<div class="section level2">
<h2 id="optimizing-the-effect-prior-hyperparameters">Optimizing the effect prior hyperparameters<a class="anchor" aria-label="anchor" href="#optimizing-the-effect-prior-hyperparameters"></a>
</h2>
<p>We recommend optimizing the effect prior hyperparmeters by an
empirical Bayes approach. In this approach, we obtain the <span class="math inline">\(\phi_g\)</span>, <span class="math inline">\(\phi_t\)</span>, and <span class="math inline">\(\phi_{g \times t}\)</span> hyperparameter values
that maximize the sum of <span class="math inline">\(\log\)</span>-transformed marginal likelihood
across all feature-SNP pairs that are being analyzed. The most
conceptually straightforward method is to perform a grid search (see <a href="https://yharigaya.github.io/classifygxt/#publication" class="external-link">our
manuscript</a> for details). We recommend using a workflow such as <a href="https://snakemake.readthedocs.io/en/stable/" class="external-link">Snakemake</a>. For
faster computation, we use MAP estimation and Laplace approximation,
setting <code>method="map.lap"</code> (see <a href="https://yharigaya.github.io/classifygxt/articles/classifygxt.html#performing-bms" class="external-link">Performing
BMS</a> above). From the BMS result for a feature-SNP pair, the <span class="math inline">\(\log\)</span> of the marginal likelihood can be
extracted as follows.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">ln.p.y</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">ln.p.y</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] -257.3321</span></span></code></pre></div>
<p>Since error bounds for Laplace approximation are not easily obtained,
we recommend rerunning MCMC and bridge sampling with optimal
hyperparameter values for the final result.</p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.1.2 (2021-11-01)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Big Sur 10.16</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib</span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] classifygxt_0.1.0 ggplot2_3.4.4    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] matrixStats_0.61.0      fs_1.5.0                fontquiver_0.2.1       </span></span>
<span><span class="co">#&gt;  [4] rstan_2.21.5            tools_4.1.2             bslib_0.3.0            </span></span>
<span><span class="co">#&gt;  [7] utf8_1.2.2              R6_2.5.1                DBI_1.1.1              </span></span>
<span><span class="co">#&gt; [10] colorspace_2.0-2        withr_3.0.0             tidyselect_1.1.1       </span></span>
<span><span class="co">#&gt; [13] gridExtra_2.3           prettyunits_1.1.1       processx_3.8.1         </span></span>
<span><span class="co">#&gt; [16] Brobdingnag_1.2-7       curl_4.3.2              compiler_4.1.2         </span></span>
<span><span class="co">#&gt; [19] extrafontdb_1.0         textshaping_0.3.6       cli_3.6.1              </span></span>
<span><span class="co">#&gt; [22] desc_1.4.3              fontBitstreamVera_0.1.1 labeling_0.4.2         </span></span>
<span><span class="co">#&gt; [25] sass_0.4.6              scales_1.3.0            mvtnorm_1.1-3          </span></span>
<span><span class="co">#&gt; [28] callr_3.7.6             pkgdown_2.0.9           systemfonts_1.0.4      </span></span>
<span><span class="co">#&gt; [31] stringr_1.4.0           digest_0.6.27           StanHeaders_2.21.0-7   </span></span>
<span><span class="co">#&gt; [34] rmarkdown_2.11          gfonts_0.2.0            pkgconfig_2.0.3        </span></span>
<span><span class="co">#&gt; [37] htmltools_0.5.5         extrafont_0.19          highr_0.9              </span></span>
<span><span class="co">#&gt; [40] fastmap_1.1.1           htmlwidgets_1.5.4       rlang_1.1.1            </span></span>
<span><span class="co">#&gt; [43] rstudioapi_0.13         httpcode_0.3.0          shiny_1.7.1            </span></span>
<span><span class="co">#&gt; [46] farver_2.1.0            jquerylib_0.1.4         generics_0.1.2         </span></span>
<span><span class="co">#&gt; [49] jsonlite_1.7.2          dplyr_1.0.10            inline_0.3.19          </span></span>
<span><span class="co">#&gt; [52] magrittr_2.0.1          loo_2.5.1               Matrix_1.3-4           </span></span>
<span><span class="co">#&gt; [55] Rcpp_1.0.7              munsell_0.5.0           fansi_0.5.0            </span></span>
<span><span class="co">#&gt; [58] gdtools_0.3.3           viridis_0.6.1           lifecycle_1.0.4        </span></span>
<span><span class="co">#&gt; [61] stringi_1.7.4           yaml_2.2.1              pkgbuild_1.2.0         </span></span>
<span><span class="co">#&gt; [64] grid_4.1.2              hrbrthemes_0.8.0        parallel_4.1.2         </span></span>
<span><span class="co">#&gt; [67] promises_1.2.0.1        crayon_1.5.1            lattice_0.20-45        </span></span>
<span><span class="co">#&gt; [70] knitr_1.34              ps_1.6.0                pillar_1.9.0           </span></span>
<span><span class="co">#&gt; [73] codetools_0.2-18        stats4_4.1.2            rstantools_2.4.0       </span></span>
<span><span class="co">#&gt; [76] crul_1.2.0              glue_1.6.2              evaluate_0.14          </span></span>
<span><span class="co">#&gt; [79] fontLiberation_0.1.0    RcppParallel_5.1.5      vctrs_0.6.2            </span></span>
<span><span class="co">#&gt; [82] httpuv_1.6.3            Rttf2pt1_1.3.12         gtable_0.3.0           </span></span>
<span><span class="co">#&gt; [85] purrr_1.0.1             assertthat_0.2.1        cachem_1.0.6           </span></span>
<span><span class="co">#&gt; [88] xfun_0.26               mime_0.11               xtable_1.8-4           </span></span>
<span><span class="co">#&gt; [91] coda_0.19-4             later_1.3.0             ragg_1.2.5             </span></span>
<span><span class="co">#&gt; [94] viridisLite_0.4.0       tibble_3.2.1            memoise_2.0.1          </span></span>
<span><span class="co">#&gt; [97] ellipsis_0.3.2          bridgesampling_1.1-2</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-hipsci_consortium_shared_2018" class="csl-entry">
HIPSCI Consortium, Kaur Alasoo, Julia Rodrigues, Subhankar Mukhopadhyay,
Andrew J. Knights, Alice L. Mann, Kousik Kundu, Christine Hale, Gordon
Dougan, and Daniel J. Gaffney. 2018. <span>“Shared Genetic Effects on
Chromatin and Gene Expression Indicate a Role for Enhancer Priming in
Immune Response.”</span> <em>Nature Genetics</em> 50 (3): 424–31. <a href="https://doi.org/10.1038/s41588-018-0046-7" class="external-link">https://doi.org/10.1038/s41588-018-0046-7</a>.
</div>
<div id="ref-matoba_stimulating_2024" class="csl-entry">
Matoba, Nana, Brandon D. Le, Jordan M. Valone, Justin M. Wolter, Jessica
T. Mory, Dan Liang, Nil Aygün, et al. 2024. <span>“Stimulating
<span>Wnt</span> Signaling Reveals Context-Dependent Genetic Effects on
Gene Regulation in Primary Human Neural Progenitors.”</span> <em>Nature
Neuroscience</em>, September. <a href="https://doi.org/10.1038/s41593-024-01773-6" class="external-link">https://doi.org/10.1038/s41593-024-01773-6</a>.
</div>
<div id="ref-palowitch_estimation_2018" class="csl-entry">
Palowitch, John, Andrey Shabalin, Yi-Hui Zhou, Andrew B. Nobel, and Fred
A. Wright. 2018. <span>“Estimation of Cis-<span class="nocase">eQTL</span> Effect Sizes Using a Log of Linear Model:
<span>Effect</span> <span>Sizes</span> of Cis-<span class="nocase">eQTLs</span> <span>Using</span> a <span>Log</span> of
<span>Linear</span> <span>Model</span>.”</span> <em>Biometrics</em> 74
(2): 616–25. <a href="https://doi.org/10.1111/biom.12810" class="external-link">https://doi.org/10.1111/biom.12810</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yuriko Harigaya, Michael Love, William Valdar.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
