<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using ClassifyGxT with TensorQTL â€¢ classifygxt</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using ClassifyGxT with TensorQTL">
<meta property="og:description" content="classifygxt">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">classifygxt</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/classifygxt.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/tensorqtl.html">Using ClassifyGxT with TensorQTL</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/yharigaya/classifygxt/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using ClassifyGxT with TensorQTL</h1>
                        <h4 data-toc-skip class="author">Yuriko
Harigaya, Michael Love, William Valdar</h4>
            
            <h4 data-toc-skip class="date">2024-11-22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/yharigaya/classifygxt/blob/HEAD/vignettes/tensorqtl.Rmd" class="external-link"><code>vignettes/tensorqtl.Rmd</code></a></small>
      <div class="hidden name"><code>tensorqtl.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette demonstrates how to format input data for ClassifyGxT
from example input/output data for a widely used program for expression
QTL mapping, <a href="https://github.com/broadinstitute/tensorqtl" class="external-link">TensorQTL</a>. The
procedure described here focuses on data without repeated measurements
per donor (e.g., gene expression data from clinical samples). For
analyzing data with repeated measurements (e.g., molecular phenotypes
from <em>in-vitro</em> cell systems), example code can be found at <a href="https://bitbucket.org/steinlabunc/wnt-rqtls" class="external-link uri">https://bitbucket.org/steinlabunc/wnt-rqtls</a> and <a href="https://github.com/yharigaya/classifygxt-paper" class="external-link uri">https://github.com/yharigaya/classifygxt-paper</a>.</p>
<p>Since the demonstration relies on large external data, the code in
this vignette is not executed. To execute the code, it is necessary to
download <a href="https://github.com/broadinstitute/tensorqtl/tree/master/example/data" class="external-link">example
data</a> from the TensorQTL website and store them locally. Execution
also requires installation of <a href="https://github.com/broadinstitute/tensorqtl" class="external-link">TensorQTL</a> and <a href="https://www.cog-genomics.org/plink/2.0/" class="external-link">PLINK2</a>.</p>
</div>
<div class="section level2">
<h2 id="data-preprocessing">Data preprocessing<a class="anchor" aria-label="anchor" href="#data-preprocessing"></a>
</h2>
<p>We recommend processing molecular count data as illustrated in steps
1-6 of <a href="https://yharigaya.github.io/classifygxt/articles/classifygxt.html#data-preprocessing" class="external-link">Data
preprocessing</a> prior to running TensorQTL.
<!-- If linear regression is used for BMS, variance-stabilizing transformations other than $\log$ transformation (e.g., VST from the DESeq2 package) can be used. --></p>
</div>
<div class="section level2">
<h2 id="running-tensorqtl">Running TensorQTL<a class="anchor" aria-label="anchor" href="#running-tensorqtl"></a>
</h2>
<p>In this section, we briefly illustrate how to run interaction QTL
mapping using TensorQTL prior to running ClassifyGxT.
<!-- Note that it is also possible to run interaction QTL mapping in two steps as in (refs). (Is it possible to pass a list of feature-SNP pairs?) -->
See the instruction on <a href="https://github.com/broadinstitute/tensorqtl" class="external-link">TensorQTL</a> for
details.</p>
<div class="section level3">
<h3 id="interaction-qtl-mapping-in-general">Interaction QTL mapping in general<a class="anchor" aria-label="anchor" href="#interaction-qtl-mapping-in-general"></a>
</h3>
<p>We use the following input data for performing interaction QTL
mapping using TensorQTL in a way that is compatible with
ClassifyGxT.</p>
<ul>
<li>PLINK2 genotype files (<code>.pgen</code>, <code>.psam</code>,
<code>.pvar</code>)</li>
<li>Molecular phenotypes file: A tab-delimited file of processed gene
expression data (or other molecular phenotypes) with rows and columns
representing gene (or feature) IDs and sample IDs, respectively.</li>
<li>Covariates file: A tab-delimited file of covariates with rows and
columns representing covariate names and sample IDs, respectively. The
covariates include molecular phenotype PCs (from <a href="https://yharigaya.github.io/classifygxt/articles/classifygxt.html#data-preprocessing" class="external-link">step
6</a>) and, optionally, other nuisance factors (see <a href="https://yharigaya.github.io/classifygxt/articles/classifygxt.html#data-preprocessing" class="external-link">step
5</a>).</li>
<li>Interactions file: A tab-delimited file of the independent variable
of interest (e.g., treatment conditions coded as <span class="math inline">\({1, 2}\)</span>). The first and second columns
must contain sample IDs and the treatment condition variable,
respectively. A header is not needed.</li>
</ul>
<p>In general, the following code, taken from the TensorQTL instruction,
performs interaction mapping.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">python3</span> <span class="at">-m</span> tensorqtl <span class="va">${plink_prefix_path}</span> <span class="va">${expression_bed}</span> <span class="va">${prefix}</span> <span class="dt">\</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">--covariates</span> <span class="va">${covariates_file}</span> <span class="dt">\</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">--interaction</span> <span class="va">${interactions_file}</span> <span class="dt">\</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--best_only</span> <span class="dt">\</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--mode</span> cis_nominal</span></code></pre></div>
<p>The code will generate an output file,
<code>${prefix}.cis_qtl_top_assoc.txt.gz</code>. The
<code>plink_prefix_path</code> variable must be the path to the PLINK2
genotype files. That is, the paths to the files must be
<code>${plink_prefix_path}.pgen</code>,
<code>${plink_prefix_path}.pvar</code>, and
<code>${plink_prefix_path}.psam</code>. The <code>expression_bed</code>,
<code>covariates_file</code>, and <code>interactions_file</code>
variables must be the paths to the molecular phenotypes, covariates, and
interactions files, respectively. With this particular setting, nominal
<span class="math inline">\(P\)</span> values smaller than 0.01 will be
included in the output file.</p>
</div>
<div class="section level3">
<h3 id="an-example-of-interaction-qtl-mapping">An example of interaction QTL mapping<a class="anchor" aria-label="anchor" href="#an-example-of-interaction-qtl-mapping"></a>
</h3>
<p>For concreteness, we illustrate the procedure using <a href="https://github.com/broadinstitute/tensorqtl/tree/master/example/data" class="external-link">example
datasets</a> available from the TensorQTL website.</p>
<ul>
<li><code>GEUVADIS.445_samples.GRCh38.20170504.maf01.filtered.nodup.chr18.pgen</code></li>
<li><code>GEUVADIS.445_samples.GRCh38.20170504.maf01.filtered.nodup.chr18.psam</code></li>
<li><code>GEUVADIS.445_samples.GRCh38.20170504.maf01.filtered.nodup.chr18.pvar</code></li>
<li><code>GEUVADIS.445_samples.covariates.txt</code></li>
<li><code>GEUVADIS.445_samples.expression.bed.gz</code></li>
</ul>
<p>We assume that the above files are stored in the <code>input/</code>
directory under the working directory. We will save intermediate files
in a separate directory, <code>processed/</code>. The covariate file
(<code>GEUVADIS.445_samples.covariates.txt</code>) contains 26
covariates including the sex variable (coded as 1 and 2). Here, we
attempt to map QTLs with gene-by-sex interactions. For this purpose, we
first create an interaction file, which contains the sample names and
the sex variable. In R, the following code can be used.
<!-- The `in.dir` variable specifies the directory where the input files are stored.
The `out.dir` variable specifies the directory where the output files will be written. --></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># read in the data</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>covar.file <span class="ot">&lt;-</span> <span class="st">"input/GEUVADIS.445_samples.covariates.txt"</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>covar <span class="ot">&lt;-</span> covar.file <span class="sc">%&gt;%</span> <span class="fu">read.delim</span>(<span class="at">row.names=</span><span class="dv">1</span>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co"># create a data frame</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>sample <span class="ot">&lt;-</span> <span class="fu">colnames</span>(covar)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>value <span class="ot">&lt;-</span> covar[<span class="dv">26</span>, ] <span class="sc">%&gt;%</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>    unlist <span class="sc">%&gt;%</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    unname <span class="sc">%&gt;%</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>    as.integer</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>int <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">sample=</span>sample, <span class="at">value=</span>value)</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a><span class="co"># specify the name of an interactions file</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>int.file <span class="ot">&lt;-</span> <span class="st">"processed/GEUVADIS.445_samples.interactions.txt"</span><span class="er">)</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a><span class="co"># write the sex variable to an file</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>    int, <span class="at">file=</span>int.file, <span class="at">quote=</span><span class="cn">FALSE</span>, <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">row.names=</span><span class="cn">FALSE</span>, <span class="at">col.names=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<p>We next create a new covariate file by removing the sex variable from
the covariate file.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">covar2</span> <span class="op">&lt;-</span> <span class="va">covar</span><span class="op">[</span><span class="op">-</span><span class="fl">26</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># specify the name of a new covariates file</span></span>
<span><span class="va">covar2.file</span> <span class="op">&lt;-</span> <span class="st">"processed/GEUVADIS.445_samples.covariates2.txt"</span></span>
<span></span>
<span><span class="co"># write to the covariates to a file</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span></span>
<span>    <span class="va">covar2</span>, file<span class="op">=</span><span class="va">covar2.file</span>, quote<span class="op">=</span><span class="cn">FALSE</span>, sep<span class="op">=</span><span class="st">"\t"</span>,</span>
<span>    row.names<span class="op">=</span><span class="cn">TRUE</span>, col.names<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>We then perform interaction QTL mapping by typing the following code
in a command-line application.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># specify the input file names</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="va">plink_prefix_path</span><span class="op">=</span><span class="st">"input/GEUVADIS.445_samples.GRCh38.20170504.maf01.filtered.nodup.chr18"</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="va">expression_bed</span><span class="op">=</span><span class="st">"input/GEUVADIS.445_samples.expression.bed.gz"</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="va">covariates_file</span><span class="op">=</span><span class="st">"processed/GEUVADIS.445_samples.covariates2.txt"</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="va">interactions_file</span><span class="op">=</span><span class="st">"processed/GEUVADIS.445_samples.interactions.txt"</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="co"># specify an output prefix</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="va">prefix</span><span class="op">=</span><span class="st">"processed/GEUVADIS.445_samples"</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="ex">python3</span> <span class="at">-m</span> tensorqtl <span class="va">${plink_prefix_path}</span> <span class="va">${expression_bed}</span> <span class="va">${prefix}</span> <span class="dt">\</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">--covariates</span> <span class="va">${covariates_file}</span> <span class="dt">\</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">--interaction</span> <span class="va">${interactions_file}</span> <span class="dt">\</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">--best_only</span> <span class="dt">\</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">--mode</span> cis_nominal</span></code></pre></div>
<p>The result will be saved as
<code>path/to/GEUVADIS.445_samples.cis_qtl_top_assoc.txt.gz</code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="formatting-data-for-classifygxt">Formatting data for ClassifyGxT<a class="anchor" aria-label="anchor" href="#formatting-data-for-classifygxt"></a>
</h2>
<p>To analyze the interaction QTLs using ClassifyGxT, we first convert
the PLINK2 genotype files to a <code>.traw</code> file by typing the
following code in a command-line application.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="va">out_prefix</span><span class="op">=</span><span class="st">"processed/GEUVADIS.445_samples.GRCh38.20170504.maf01.filtered.nodup.chr18"</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="ex">plink2</span> <span class="at">--pfile</span> <span class="va">${plink_prefix_path}</span> <span class="at">--export</span> Av <span class="at">--out</span> <span class="va">${out_prefix}</span></span></code></pre></div>
<!-- We also extract IDs of genes and SNPs with significant GxT interactions from the output from TensorQTL. -->
<!-- We then preprocess the expression matrix according to `Get started`. -->
<p>Next, in R, we specify the input file names.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qtl.file</span> <span class="op">&lt;-</span> <span class="st">"processed/GEUVADIS.445_samples.chr18.cis_qtl_top_assoc.txt.gz"</span></span>
<span><span class="va">geno.file</span> <span class="op">&lt;-</span> <span class="st">"processed/GEUVADIS.445_samples.GRCh38.20170504.maf01.filtered.nodup.chr18.traw"</span></span>
<span><span class="va">pheno.file</span> <span class="op">&lt;-</span> <span class="st">"input/GEUVADIS.445_samples.expression.bed.gz"</span></span>
<span><span class="va">covar.file</span> <span class="op">&lt;-</span> <span class="st">"processed/GEUVADIS.445_samples.covariates2.txt"</span></span>
<span><span class="va">int.file</span> <span class="op">&lt;-</span> <span class="st">"processed/GEUVADIS.445_samples.interactions.txt"</span></span></code></pre></div>
<p>We then attach <em>classifygxt</em> and obtain a list object using
<code><a href="../reference/format_input.html">format_input()</a></code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/yharigaya/classifygxt" class="external-link">classifygxt</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">input.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/format_input.html">format_input</a></span><span class="op">(</span></span>
<span>    qtl<span class="op">=</span><span class="va">qtl.file</span>, geno<span class="op">=</span><span class="va">geno.file</span>, pheno<span class="op">=</span><span class="va">pheno.file</span>,</span>
<span>    covar<span class="op">=</span><span class="va">covar.file</span>, int<span class="op">=</span><span class="va">int.file</span><span class="op">)</span></span></code></pre></div>
<p>The <code>input.list</code> object above is a list of lists, each of
which corresponds to a feature-SNP pair and can be used as input for
<code><a href="../reference/do_bms.html">do_bms()</a></code>. Depending on the format of the input files, it
may be necessary to specify additional input arguments. See the function
documentation (<code><a href="../reference/format_input.html">?format_input</a></code>) for details.</p>
<!-- comment on the column numbers -->
<!-- TO-DO (optional)
- create a dummy txt file (feature, snp)
- write a function to process the data (regress, pca, residualize)
-->
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.1.2 (2021-11-01)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Big Sur 10.16</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib</span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] rstudioapi_0.13   knitr_1.34        magrittr_2.0.1    R6_2.5.1         </span></span>
<span><span class="co">#&gt;  [5] ragg_1.2.5        rlang_1.1.1       fastmap_1.1.1     stringr_1.4.0    </span></span>
<span><span class="co">#&gt;  [9] tools_4.1.2       xfun_0.26         cli_3.6.1         jquerylib_0.1.4  </span></span>
<span><span class="co">#&gt; [13] htmltools_0.5.5   systemfonts_1.0.4 yaml_2.2.1        digest_0.6.27    </span></span>
<span><span class="co">#&gt; [17] lifecycle_1.0.4   pkgdown_2.0.9     textshaping_0.3.6 purrr_1.0.1      </span></span>
<span><span class="co">#&gt; [21] sass_0.4.6        htmlwidgets_1.5.4 vctrs_0.6.2       fs_1.5.0         </span></span>
<span><span class="co">#&gt; [25] memoise_2.0.1     cachem_1.0.6      evaluate_0.14     rmarkdown_2.11   </span></span>
<span><span class="co">#&gt; [29] stringi_1.7.4     compiler_4.1.2    bslib_0.3.0       desc_1.4.3       </span></span>
<span><span class="co">#&gt; [33] jsonlite_1.7.2</span></span></code></pre></div>
<!-- # References -->
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yuriko Harigaya, Michael Love, William Valdar.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
