<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Format input — format_input • classifygxt</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Format input — format_input"><meta property="og:description" content="This function takes input and output files of
TensorQTL and
returns a list object that can be used as input for
do_bms. It currenly only handles completely unpaired data,
where there is no repeated measurement per donor (subject). This
includes typical molecular data from clinical trials."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">classifygxt</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/classifygxt.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/tensorqtl.html">Using ClassifyGxT with TensorQTL</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/yharigaya/classifygxt/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Format input</h1>
    <small class="dont-index">Source: <a href="https://github.com/yharigaya/classifygxt/blob/HEAD/R/utils.R" class="external-link"><code>R/utils.R</code></a></small>
    <div class="hidden name"><code>format_input.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function takes input and output files of
<a href="https://github.com/broadinstitute/tensorqtl" class="external-link">TensorQTL</a> and
returns a list object that can be used as input for
<code><a href="do_bms.html">do_bms</a></code>. It currenly only handles completely unpaired data,
where there is no repeated measurement per donor (subject). This
includes typical molecular data from clinical trials.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">format_input</span><span class="op">(</span></span>
<span>  <span class="va">qtl</span>,</span>
<span>  <span class="va">pheno</span>,</span>
<span>  <span class="va">geno</span>,</span>
<span>  <span class="va">covar</span>,</span>
<span>  <span class="va">int</span>,</span>
<span>  pheno.col <span class="op">=</span> <span class="st">"phenotype_id"</span>,</span>
<span>  variant.col <span class="op">=</span> <span class="st">"variant_id"</span>,</span>
<span>  feat.col <span class="op">=</span> <span class="st">"gene_id"</span>,</span>
<span>  snp.col <span class="op">=</span> <span class="st">"SNP"</span>,</span>
<span>  pheno.rm <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,</span>
<span>  geno.rm <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">6</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>qtl</dt>
<dd><p>A character string specfying the name of the output file
from interacton QTL mapping from TensorQTL.</p></dd>


<dt>pheno</dt>
<dd><p>A character string specifynig the name of the file
containing molecular phenotes. This must be the same as the
input file used for TensorQTL.</p></dd>


<dt>geno</dt>
<dd><p>A character string specifying the name of the genotype
file in <a href="https://www.cog-genomics.org/plink/2.0" class="external-link">PLINK2</a>
<code>.traw</code> format.</p></dd>


<dt>covar</dt>
<dd><p>A character string specifyng the name of the file
containing covariates. This must be the same as the input file
used for TensorQTL.</p></dd>


<dt>int</dt>
<dd><p>A character string specifying the name of the file
containing the condition variable of interest, coded as 1,
2. This must be the same as the input file used for TensorQTL.</p></dd>


<dt>pheno.col</dt>
<dd><p>A character string specifying the name of the
column of <code>qtl</code> containing the feature IDs.</p></dd>


<dt>variant.col</dt>
<dd><p>A character string specifying the name of the
column of <code>qtl</code> containing the feature IDs.</p></dd>


<dt>feat.col</dt>
<dd><p>A character string specifying the name of the
column of <code>pheno</code> containing the feature IDs.</p></dd>


<dt>snp.col</dt>
<dd><p>A character string specifying the name of the
column of <code>geno</code> containing the SNP IDs.</p></dd>


<dt>pheno.rm</dt>
<dd><p>A vector of integers corresponding to the column
numbers for non-phenotype entries in <code>pheno</code>.</p></dd>


<dt>geno.rm</dt>
<dd><p>A vector of integers corresponding to the column
numbers for non-phenotype entries in <code>geno</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list of lists containing:</p><ul><li><p><code>y</code> - A vector of phenotypes.</p></li>
<li><p><code>g</code> - A vector of genotypes.</p></li>
<li><p><code>t</code> - A vector of treatment indicators.</p></li>
<li><p><code>subject</code> - A vector of subject.</p></li>
<li><p><code>feat.id</code> - A character string spcifyng the feature ID.</p></li>
<li><p><code>snp.id</code> - A character string specifying the SNP ID.</p></li>
</ul></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yuriko Harigaya, Michael Love, William Valdar.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

